src/reactor.rs

```diff
@@ -0,0 +1,741 @@
+        request: Request,
+    ) where
+        K: Keychain,
+        M: Moderator,
+    {
+        let response = match request {
+            Request::Commit { id, height, round } => Response::Commit {
+                id,
+                committed: self.commit(moderator, height, round).await,
+            },
+
+            Request::Identity { id, height } => Response::Identity {
+                id,
+                public: keychain
+                    .public(height)
+                    .ok()
+                    .flatten()
+                    .map(|k| k.into_owned()),
+            },
+
+            Request::Initialize {
+                id,
+                start,
+                validity,
+            } => Response::Initialize {
+                id,
+                initialized: keychain
+                    .public(start)
+                    .ok()
+                    .flatten()
+                    .map(|public| self.add_validator(public.into_owned(), start, validity))
+                    .is_some(),
+            },
+
+            Request::Round { id } => {
+                let height = self.height();
+                let round = self.round(moderator.now());
+                let leader = self.leader(round).copied().unwrap_or_default();
+                let public = keychain
+                    .public(height)
+                    .unwrap_or_default()
+                    .map(|p| p.into_owned())
+                    .unwrap_or_default();
+                let step = self.validator_step(height, round, &public);
+
+                Response::Round {
+                    id,
+                    height,
+                    round,
+                    leader,
+                    step,
+                }
+            }
+        };
+
+        let response = Message::Response(response);
+
+        moderator.send(response, self.timeout).await;
+    }
+
+    /// Receive a new message, mutating the internal state
+    pub async fn receive<K, M>(&mut self, keychain: &K, moderator: &mut M, message: Message)
+    where
+        K: Keychain,
+        M: Moderator,
+    {
+        #[cfg(feature = "trace")]
+        tracing::trace!("receiving message {:?}", message);
+
+        match message {
+            Message::Event(e) => self.receive_event(e).await,
+            Message::Notification(n) => self.receive_notification(keychain, moderator, n).await,
+            Message::Request(r) => self.receive_request(keychain, moderator, r).await,
+            Message::Response(_) => (),
+        }
+    }
+
+    /// Check the current status of the reactor, producing an event, if applicable
+    pub async fn heartbeat<K, M>(&mut self, keychain: &K, moderator: &mut M) -> Result<(), Error>
+    where
+        K: Keychain,
+        M: Moderator,
+    {
+        let height = self.height();
+
+        #[cfg(feature = "trace")]
+        tracing::trace!("heartbeat height {}", height);
+
+        // If no public key is available in the keychain, the node is idle
```

Comment:
No consensus is possible if we don't have public keys. The intent of the consensus is to spawn new votes to the network; that will never happen if we are not aware of the peer validators

Reply:
The keychain here contains a private key for your validator, it doesn't contain all validator public keys/stakes. I concluded that because of these lines: https://github.com/FuelLabs/fuel-bft/blob/7acc238ee02ec8cbc726deb97519b075fc0ab538/src/tokio_reactor.rs#L103-L106
Not sure where full validator set is stored but in future, they will be obtained from `fuel-relayer`.

And you still want to check consensus even if you are not participating in it.

Reply:
Captured here: https://github.com/FuelLabs/fuel-bft/issues/16